tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/sort.h>
+
+#include "tempesta_fw.h"
+#include "apm.h"
+#include "log.h"
+#include "server.h"
+
+MODULE_AUTHOR(TFW_AUTHOR);
+MODULE_DESCRIPTION("Tempesta Ratio Scheduler");
+MODULE_VERSION("0.1.0");
+MODULE_LICENSE("GPL");
+
+#define TFW_SCHED_RATIO_INTVL	(HZ / 20)	/* The timer periodicity. */
+
+/**
+ * Individual upstream server descriptor.
+ *
+ * Connections may go up or down during failover process.
+ * Only fully established connections are considered by scheduler.
+ *
+ * @srv		- pointer to server structure.
+ * @conns	- list of pointers to server connection structures.
+ * @counter	- monotonic counter for choosing the next connection.
+ * @conn_n	- number of connections to server.
+ * @seq		- current sequence number for APM stats.
+ */
+typedef struct {
+	TfwServer	*srv;
+	TfwSrvConn	**conns;
+	atomic64_t	counter;
+	size_t		conn_n;
+	unsigned int	seq;
+} TfwRatioSrvDesc;
+
+/**
+ * Server data for scheduler.
+ *
+ * @sdidx	- index of server descriptor this data is for.
+ * @weight	- server weight.
+ * @cratio	- current server ratio.
+ * @oratio	- original server ratio.
+ */
+typedef struct {
+	size_t		sdidx;
+	unsigned int	weight;
+	unsigned int	cratio;
+	unsigned int	oratio;
+} TfwRatioSrvData;
+
+/**
+ * Scheduler iteration data.
+ *
+ * @lock	- must be in the same cache line for faster operations.
+ * @csidx	- index of current server data entry.
+ * @reidx	- index of next server data entry which ratio we need
+ *		  to reset, or @srv_n if no resetting is needed.
+ * @riter	- ratio iteration, indicates the number of times we need
+ *		  to choose all servers before the current one until we
+ *		  can choose the current server.
+ * @crsum	- current sum of all ratios, used to avoid scanning the
+ *		  list of servers with fully zeroed ratios.
+ * @orsum	- original sum of all ratios, used to reset @crsum.
+ */
+typedef struct {
+	spinlock_t	lock;
+	size_t		csidx;
+	size_t		reidx;
+	unsigned int	riter;
+	unsigned long	crsum;
+	unsigned long	orsum;
+} TfwRatioSchData;
+
+/**
+ * Historic (past) data unit for an individual upstream server.
+ *
+ * @x		- count of timer function invocations.
+ * @y		- RTT from APM in msecs.
```

Comment:
The primary reason was that derivative members become too long, like, say, `cnt_avg_rtt_avg`, etc. In calculations these names are used multiple times in the same statement which adds up to very long statements that are difficult to read and leads to ugly formatting. I believe the comments here are sufficient to understand the actual meaning of the calculations, and the algorithm can be easily correlated with the original in `slr.c`.